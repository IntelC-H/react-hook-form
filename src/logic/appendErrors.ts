import {
  InternalFieldName,
  ValidateResult,
  InternalFieldErrors,
} from '../types';

export default <TFieldValues>(
  name: InternalFieldName<TFieldValues>,
  validateAllFieldCriteria: boolean,
  errors: InternalFieldErrors<TFieldValues>,
  type: string,
  message: ValidateResult,
) =>
  validateAllFieldCriteria
    ? {
        ...errors[name],
        types: {
          ...(errors[name] && errors[name]!.types ? errors[name]!.types : {}),
          [type]: message || true,
        },
      }
    : {};
