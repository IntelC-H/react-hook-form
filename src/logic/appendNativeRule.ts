export default function appendNativeRule(ref, data) {
  Object.entries(data).forEach(([key, value]) => {
    if (key === 'required') {
      ref[key] = true;
    } else if (key === 'pattern' && value instanceof RegExp) {
      ref[key] = value.source;
    } else {
      ref[key] = value;
    }
  });
}
