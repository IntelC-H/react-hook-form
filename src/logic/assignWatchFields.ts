import combineFieldValues from './combineFieldValues';
import get from '../utils/get';
import getPath from '../utils/getPath';
import isEmptyObject from '../utils/isEmptyObject';
import isUndefined from '../utils/isUndefined';
import { FieldValue, FieldValues, FieldName } from '../types';

export default <FormValues extends FieldValues>(
  fieldValues: FormValues,
  fieldName: FieldName<FormValues>,
  watchFields: Set<FieldName<FormValues>>,
): FieldValue<FormValues> | Partial<FormValues> | undefined => {
  if (isEmptyObject(fieldValues)) {
    return undefined;
  }

  if (!isUndefined(fieldValues[fieldName])) {
    watchFields.add(fieldName);
    return fieldValues[fieldName];
  }

  const values = get(combineFieldValues(fieldValues), fieldName);

  if (!isUndefined(values)) {
    getPath<FormValues>(fieldName, values).forEach(name =>
      watchFields.add(name),
    );
  }

  return values;
};
