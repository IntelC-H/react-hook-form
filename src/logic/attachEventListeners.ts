import { Field } from '../type';
import isCheckbox from '../utils/isCheckBox';

export default function attachEventListeners({
  field,
  validateAndStateUpdate,
  isRadio,
}: {
  field: Field;
  isRadio: boolean;
  validateAndStateUpdate: (any) => void;
}): void {
  const { ref } = field;
  ref.addEventListener(isCheckbox(ref.type) || isRadio ? 'change' : 'input', validateAndStateUpdate);
  ref.addEventListener('blur', validateAndStateUpdate);
}
