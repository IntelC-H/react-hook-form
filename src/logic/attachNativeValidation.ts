import isRegex from '../utils/isRegex';
import getValueAndMessage from './getValueAndMessage';
import { INPUT_VALIDATION_RULES } from '../constants';
import { FieldValues, Ref } from '../types';

export default function attachNativeValidation(
  ref: Ref,
  rules: FieldValues,
): void {
  Object.entries(rules).forEach(([key, ruleValue]) => {
    const { value } = getValueAndMessage(ruleValue);

    if (key === INPUT_VALIDATION_RULES.pattern && isRegex(value)) {
      ref[key] = value.source;
    } else {
      ref[key] = key === INPUT_VALIDATION_RULES.pattern || value;
    }
  });
}
