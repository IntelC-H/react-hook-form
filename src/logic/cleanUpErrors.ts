import { ErrorMessages } from '../types';

export default <Data>(errors: ErrorMessages<Data>, fields: string[]) =>
  Object.entries(errors).reduce(
    (result, [field, err]) =>
      fields.some(f => f === field) ? result : { ...result, [field]: err },
    {},
  );
