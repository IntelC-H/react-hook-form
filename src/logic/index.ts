export * from './appendErrors';
export * from './transformToNestObject';
