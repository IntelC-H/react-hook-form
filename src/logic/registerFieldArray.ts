import * as React from 'react';
import set from '../utils/set';
import { InternalFieldName } from '../types';

export const registerFieldArray = <T extends Object[]>(
  fieldsRef: React.MutableRefObject<any>,
  name: InternalFieldName,
  values: T,
  index: number = 0,
  parentName = '',
) => {
  Array.isArray(values) &&
    values.forEach((appendValueItem, valueIndex) =>
      Object.entries(appendValueItem).forEach((fieldNameAndValue) => {
        if (fieldNameAndValue) {
          const [key, value] = fieldNameAndValue;
          const inputName = `${parentName || name}.${
            parentName ? valueIndex : index + valueIndex
          }.${key}`;

          Array.isArray(value)
            ? registerFieldArray(fieldsRef, name, value, valueIndex, inputName)
            : set(fieldsRef.current, inputName, {
                _f: {
                  ref: {
                    name: inputName,
                  },
                  name: inputName,
                  value,
                },
              });
        }
      }),
    );
};
