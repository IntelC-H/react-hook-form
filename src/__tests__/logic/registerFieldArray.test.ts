import { registerFieldArray } from '../../logic/registerFieldArray';

describe('registerFieldArray', () => {
  it('should register the correct field array inputs', () => {
    const fields = { current: {} };
    registerFieldArray(
      fields,
      'test',
      [
        {
          test1: '1',
          nested: [
            {
              data1: 'test1',
              data2: 'test1',
            },
          ],
        },
      ],
      0,
    );

    expect(fields.current).toMatchSnapshot();

    registerFieldArray(
      fields,
      'test',
      [
        {
          test2: '2',
          nested: [
            {
              data2: 'test2',
              data3: 'test2',
            },
            {
              data4: 'test2',
              data5: 'test2',
            },
          ],
        },
      ],
      0,
    );

    expect(fields.current).toMatchSnapshot();

    registerFieldArray(
      fields,
      'test',
      [
        {
          test: '3',
          nested: [
            {
              data3: 'test3',
              data4: 'test3',
            },
            {
              data5: 'test3',
              data6: 'test3',
            },
          ],
        },
      ],
      1,
    );

    expect(fields.current).toMatchSnapshot();

    registerFieldArray(
      fields,
      'test',
      [
        {
          test: '4',
          nested: [
            {
              data8: 'test3',
              deeper: [
                {
                  deeper1: 'test3',
                  deeper2: 'test3',
                },
                {
                  deeper3: 'test3',
                  deeper4: 'test3',
                },
              ],
            },
            {
              data10: 'test3',
              data11: {
                extra: 'test',
              },
            },
          ],
        },
      ],
      4,
    );

    expect(fields.current).toMatchSnapshot();
  });
});
