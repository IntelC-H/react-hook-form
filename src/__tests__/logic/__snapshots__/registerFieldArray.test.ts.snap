// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registerFieldArray should register the correct field array inputs 1`] = `
Object {
  "test": Array [
    Object {
      "nested": Array [
        Object {
          "data1": Object {
            "_f": Object {
              "name": "test.0.nested.0.data1",
              "ref": Object {
                "name": "test.0.nested.0.data1",
              },
              "value": "test1",
            },
          },
          "data2": Object {
            "_f": Object {
              "name": "test.0.nested.0.data2",
              "ref": Object {
                "name": "test.0.nested.0.data2",
              },
              "value": "test1",
            },
          },
        },
      ],
      "test1": Object {
        "_f": Object {
          "name": "test.0.test1",
          "ref": Object {
            "name": "test.0.test1",
          },
          "value": "1",
        },
      },
    },
  ],
}
`;

exports[`registerFieldArray should register the correct field array inputs 2`] = `
Object {
  "test": Array [
    Object {
      "nested": Array [
        Object {
          "data1": Object {
            "_f": Object {
              "name": "test.0.nested.0.data1",
              "ref": Object {
                "name": "test.0.nested.0.data1",
              },
              "value": "test1",
            },
          },
          "data2": Object {
            "_f": Object {
              "name": "test.0.nested.0.data2",
              "ref": Object {
                "name": "test.0.nested.0.data2",
              },
              "value": "test2",
            },
          },
          "data3": Object {
            "_f": Object {
              "name": "test.0.nested.0.data3",
              "ref": Object {
                "name": "test.0.nested.0.data3",
              },
              "value": "test2",
            },
          },
        },
        Object {
          "data4": Object {
            "_f": Object {
              "name": "test.0.nested.1.data4",
              "ref": Object {
                "name": "test.0.nested.1.data4",
              },
              "value": "test2",
            },
          },
          "data5": Object {
            "_f": Object {
              "name": "test.0.nested.1.data5",
              "ref": Object {
                "name": "test.0.nested.1.data5",
              },
              "value": "test2",
            },
          },
        },
      ],
      "test1": Object {
        "_f": Object {
          "name": "test.0.test1",
          "ref": Object {
            "name": "test.0.test1",
          },
          "value": "1",
        },
      },
      "test2": Object {
        "_f": Object {
          "name": "test.0.test2",
          "ref": Object {
            "name": "test.0.test2",
          },
          "value": "2",
        },
      },
    },
  ],
}
`;

exports[`registerFieldArray should register the correct field array inputs 3`] = `
Object {
  "test": Array [
    Object {
      "nested": Array [
        Object {
          "data1": Object {
            "_f": Object {
              "name": "test.0.nested.0.data1",
              "ref": Object {
                "name": "test.0.nested.0.data1",
              },
              "value": "test1",
            },
          },
          "data2": Object {
            "_f": Object {
              "name": "test.0.nested.0.data2",
              "ref": Object {
                "name": "test.0.nested.0.data2",
              },
              "value": "test2",
            },
          },
          "data3": Object {
            "_f": Object {
              "name": "test.0.nested.0.data3",
              "ref": Object {
                "name": "test.0.nested.0.data3",
              },
              "value": "test2",
            },
          },
        },
        Object {
          "data4": Object {
            "_f": Object {
              "name": "test.0.nested.1.data4",
              "ref": Object {
                "name": "test.0.nested.1.data4",
              },
              "value": "test2",
            },
          },
          "data5": Object {
            "_f": Object {
              "name": "test.0.nested.1.data5",
              "ref": Object {
                "name": "test.0.nested.1.data5",
              },
              "value": "test2",
            },
          },
        },
      ],
      "test1": Object {
        "_f": Object {
          "name": "test.0.test1",
          "ref": Object {
            "name": "test.0.test1",
          },
          "value": "1",
        },
      },
      "test2": Object {
        "_f": Object {
          "name": "test.0.test2",
          "ref": Object {
            "name": "test.0.test2",
          },
          "value": "2",
        },
      },
    },
    Object {
      "nested": Array [
        Object {
          "data3": Object {
            "_f": Object {
              "name": "test.1.nested.0.data3",
              "ref": Object {
                "name": "test.1.nested.0.data3",
              },
              "value": "test3",
            },
          },
          "data4": Object {
            "_f": Object {
              "name": "test.1.nested.0.data4",
              "ref": Object {
                "name": "test.1.nested.0.data4",
              },
              "value": "test3",
            },
          },
        },
        Object {
          "data5": Object {
            "_f": Object {
              "name": "test.1.nested.1.data5",
              "ref": Object {
                "name": "test.1.nested.1.data5",
              },
              "value": "test3",
            },
          },
          "data6": Object {
            "_f": Object {
              "name": "test.1.nested.1.data6",
              "ref": Object {
                "name": "test.1.nested.1.data6",
              },
              "value": "test3",
            },
          },
        },
      ],
      "test": Object {
        "_f": Object {
          "name": "test.1.test",
          "ref": Object {
            "name": "test.1.test",
          },
          "value": "3",
        },
      },
    },
  ],
}
`;

exports[`registerFieldArray should register the correct field array inputs 4`] = `
Object {
  "test": Array [
    Object {
      "nested": Array [
        Object {
          "data1": Object {
            "_f": Object {
              "name": "test.0.nested.0.data1",
              "ref": Object {
                "name": "test.0.nested.0.data1",
              },
              "value": "test1",
            },
          },
          "data2": Object {
            "_f": Object {
              "name": "test.0.nested.0.data2",
              "ref": Object {
                "name": "test.0.nested.0.data2",
              },
              "value": "test2",
            },
          },
          "data3": Object {
            "_f": Object {
              "name": "test.0.nested.0.data3",
              "ref": Object {
                "name": "test.0.nested.0.data3",
              },
              "value": "test2",
            },
          },
        },
        Object {
          "data4": Object {
            "_f": Object {
              "name": "test.0.nested.1.data4",
              "ref": Object {
                "name": "test.0.nested.1.data4",
              },
              "value": "test2",
            },
          },
          "data5": Object {
            "_f": Object {
              "name": "test.0.nested.1.data5",
              "ref": Object {
                "name": "test.0.nested.1.data5",
              },
              "value": "test2",
            },
          },
        },
      ],
      "test1": Object {
        "_f": Object {
          "name": "test.0.test1",
          "ref": Object {
            "name": "test.0.test1",
          },
          "value": "1",
        },
      },
      "test2": Object {
        "_f": Object {
          "name": "test.0.test2",
          "ref": Object {
            "name": "test.0.test2",
          },
          "value": "2",
        },
      },
    },
    Object {
      "nested": Array [
        Object {
          "data3": Object {
            "_f": Object {
              "name": "test.1.nested.0.data3",
              "ref": Object {
                "name": "test.1.nested.0.data3",
              },
              "value": "test3",
            },
          },
          "data4": Object {
            "_f": Object {
              "name": "test.1.nested.0.data4",
              "ref": Object {
                "name": "test.1.nested.0.data4",
              },
              "value": "test3",
            },
          },
        },
        Object {
          "data5": Object {
            "_f": Object {
              "name": "test.1.nested.1.data5",
              "ref": Object {
                "name": "test.1.nested.1.data5",
              },
              "value": "test3",
            },
          },
          "data6": Object {
            "_f": Object {
              "name": "test.1.nested.1.data6",
              "ref": Object {
                "name": "test.1.nested.1.data6",
              },
              "value": "test3",
            },
          },
        },
      ],
      "test": Object {
        "_f": Object {
          "name": "test.1.test",
          "ref": Object {
            "name": "test.1.test",
          },
          "value": "3",
        },
      },
    },
    undefined,
    undefined,
    Object {
      "nested": Array [
        Object {
          "data8": Object {
            "_f": Object {
              "name": "test.4.nested.0.data8",
              "ref": Object {
                "name": "test.4.nested.0.data8",
              },
              "value": "test3",
            },
          },
          "deeper": Array [
            Object {
              "deeper1": Object {
                "_f": Object {
                  "name": "test.4.nested.0.deeper.0.deeper1",
                  "ref": Object {
                    "name": "test.4.nested.0.deeper.0.deeper1",
                  },
                  "value": "test3",
                },
              },
              "deeper2": Object {
                "_f": Object {
                  "name": "test.4.nested.0.deeper.0.deeper2",
                  "ref": Object {
                    "name": "test.4.nested.0.deeper.0.deeper2",
                  },
                  "value": "test3",
                },
              },
            },
            Object {
              "deeper3": Object {
                "_f": Object {
                  "name": "test.4.nested.0.deeper.1.deeper3",
                  "ref": Object {
                    "name": "test.4.nested.0.deeper.1.deeper3",
                  },
                  "value": "test3",
                },
              },
              "deeper4": Object {
                "_f": Object {
                  "name": "test.4.nested.0.deeper.1.deeper4",
                  "ref": Object {
                    "name": "test.4.nested.0.deeper.1.deeper4",
                  },
                  "value": "test3",
                },
              },
            },
          ],
        },
        Object {
          "data10": Object {
            "_f": Object {
              "name": "test.4.nested.1.data10",
              "ref": Object {
                "name": "test.4.nested.1.data10",
              },
              "value": "test3",
            },
          },
          "data11": Object {
            "_f": Object {
              "name": "test.4.nested.1.data11",
              "ref": Object {
                "name": "test.4.nested.1.data11",
              },
              "value": Object {
                "extra": "test",
              },
            },
          },
        },
      ],
      "test": Object {
        "_f": Object {
          "name": "test.4.test",
          "ref": Object {
            "name": "test.4.test",
          },
          "value": "4",
        },
      },
    },
  ],
}
`;
